Userland/Libraries/LibJS/AST.h

```diff
@@ -97,6 +97,45 @@ class ASTNode : public RefCounted<ASTNode> {
     u32 m_end_offset { 0 };
 };
 
+// This is a helper class that packs an array of T after the AST node, all in the same allocation.
+template<typename Derived, typename Base, typename T>
+class ASTNodeWithTailArray : public Base {
+public:
+    template<typename... Args>
+    static NonnullRefPtr<Derived> create(size_t tail_size, SourceRange source_range, Args&&... args)
+    {
+        static_assert(alignof(Derived) % alignof(T) == 0, "need padding for tail array");
+        auto memory = ::operator new(sizeof(Derived) + tail_size * sizeof(T));
+        return adopt_ref(*::new (memory) Derived(move(source_range), static_cast<decltype(args)>(args)...));
+    }
+
+    virtual ~ASTNodeWithTailArray() override
+    {
+        for (auto& value : tail_span())
+            value.~T();
+    }
+
+    ASTNodeWithTailArray(SourceRange source_range, Span<T const> values)
+        : Base(move(source_range))
+        , m_tail_size(values.size())
+    {
+        VERIFY(values.size() <= NumericLimits<u32>::max());
+        for (size_t i = 0; i < values.size(); ++i)
+            new (&tail_data()[i]) T(values[i]);
+    }
+
+    Span<T const> tail_span() const { return { tail_data(), tail_size() }; }
+
+    T const* tail_data() const { return reinterpret_cast<T const*>(reinterpret_cast<uintptr_t>(this) + sizeof(Derived)); }
```

Comment:
Using integer arithmetic for pointer manipulation like that is technically UB (or at the very least implementation-defined). It will also cause pointer provenance issues that inhibit optimization. I would suggest downcasting `this` to `Derived` and using `+ 1` to get a pointer to the end of the object.

Reply:
I tried using your original technique of downcasting `this` to `Derived` and using `+ 1` to get a pointer to the end, but UBSAN complains about it.

Reply:
This might be because `NewExpression::create` completely sidesteps `AstNodeWithTailArray::create`, so there might be actual UB that gets hidden by the pointer-to-int cast.

I don't see how we can combine the use of CRTP with the NewExpression : CallExpression hierarchy. Assuming you don't want to refactor the class hierarchy to follow Scott Meyer's "Make non-leaf classes abstract" guideline, it should be actually fine to drop the `Derived` parameter from `AstNodeWithTailArray` and instead make it a parameter on `create()` itself. That way you can re-use it in both `CallExpression` and `NewExpression`, and `NewExpression` can even add extra data.

Reply:
I can make `Derived` a parameter on `create()`, but then how does `tail_span()` know how where to find the tail? I can make it `tail_span<Derived>()` as well, but it falls apart in the destructor where I need to walk the tail to destroy all values. (I don't know how to get `Derived` to the destructor in a nice way)

Reply:
Ah, right, of course. What about keeping everything as-is but adding `template <typename ActualDerived = Derived>` to `create()` (where we can also `static_assert` that `sizeof(ActualDerived) == sizeof(Derived)`). Then `NewExpression` can still use the proper way of construction, while the sanitizers might like the downcast.

This is really more tricky than I originally thought, sorry about that :D